### PR TITLE
Tweak timer logic to support retries

### PR DESCRIPTION
**Pull request by jesushernandez**
This pull request tries to solve the issue of tracking timings of celery tasks when there is a retry involved

### CODE REVIEW

File: celery_statsd/__init__.py [hunk 1]
```diff
@@ -17,10 +17,12 @@ def task_key(task):
     prefix = getattr(celery.current_app.conf,
                      "CELERY_STATSD_PREFIX", "celery.")
 
+    retry_number = task.request.retries if hasattr(task, 'request') else 0
+
     if isinstance(task, six.string_types):
-        return prefix + task
+        return '{}{}{}'.format(prefix, task, retry_number)
```
Comment (iserko) [inline celery_statsd/__init__.py]: this will change the task_key for every task (new or old)

**Comment by iserko** [comment]
what's the purpose of this? your description is empty

**Comment by jesushernandez** [comment]
Added description

File: celery_statsd/__init__.py (hunk 1 above)
Comment (jesushernandez) [inline celery_statsd/__init__.py]: @iserko Updated. When retries==0, it won't use any suffix

**Comment by jesushernandez** [comment]
When we have retries, the scenario would be similar to this:

```
T1 fail
    -> retry, T2, fail
    -> retry, T3, succes
```

The signals would be triggered in this order:

```
prerun(T1): pops key (emtpy), inserts key
prerun(T2): pops key, inserts key
prerun(T3): pops key, inserts key
postrun(T3): pops key (inserted by T3), sends timing
failure(T2)
postrun(T2): tries to pop key (nothing is found)
failure(T1)
postrun(T1): tries to pop key (nothing is found)
```

Which means that we will only track the time of T3 (the succesful task). Celery triggers the signal in inverse order of task execution, so the _original_ task's postrun/failure signal is sent at the end.

File: celery_statsd/__init__.py [hunk 2]
```diff
@@ -18,9 +18,9 @@ def task_key(task):
                      "CELERY_STATSD_PREFIX", "celery.")
 
     if isinstance(task, six.string_types):
-        return prefix + task
+        return '{}{}'.format(prefix, task)
```
Comment (iserko) [inline celery_statsd/__init__.py:21]: you can revert this change

File: celery_statsd/__init__.py [hunk 3]
```diff
@@ -44,16 +44,30 @@ def get_client(celery_app):
...
 
+def _pop_timer(name, group, instance):
+    try:
+        return _state.timers.pop((name, group, instance))
+    except (AttributeError, KeyError):
+        return
+
+
 def start_timer(name, group, instance):
+    _pop_timer(name, group, instance)
```
Comment (public) [inline celery_statsd/__init__.py]: Why? Won't the next line just overwrite it if it was there?

Comment (jesushernandez) [inline celery_statsd/__init__.py]: Right, removing that

File: celery_statsd/__init__.py [hunk 4]
```diff
@@ -96,7 +101,7 @@ def statsd_task_success(sender, **kwargs):
 
 
 @celery.signals.task_failure.connect
-def statsd_task_failure(sender, **kwargs):
+def statsd_task_failure(sender, task_id, **kwargs):
```
Comment (public) [inline celery_statsd/__init__.py]: Was this broken before then?

Comment (jesushernandez) [inline celery_statsd/__init__.py]: I added that while debugging, removing now

**Comment by public** [comment]
Add a test please.

**Comment by iserko** [comment]
@jesushernandez can you change the URL to the project in `__about__.py` too please?

**Comment by public** [comment]
PRs are cheap.

:+1:

**Comment by public** [comment]
Oh but you bumped the version in here. Guess we have to wait then.

**Comment by jesushernandez** [comment]
@iserko done

**Comment by iserko** [comment]
:+1:
